File: client/src/main/scala/stasis/client/encryption/secrets/DeviceSecret.scala
package stasis.client.encryption.secrets

import java.nio.charset.StandardCharsets
import java.nio.file.Path
import org.apache.pekko.stream.Materializer
import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString
import at.favre.lib.hkdf.HKDF
import stasis.client.encryption.stream.CipherStage
import stasis.core.packaging.Crate
import stasis.shared.model.devices.Device
import stasis.shared.model.users.User
import stasis.shared.secrets.SecretsConfig

import scala.concurrent.{ExecutionContext, Future}

final case class DeviceSecret(
  user: User.Id,
  device: Device.Id,
  private val secret: ByteString
)(implicit target: SecretsConfig)
    extends Secret {
  def encrypted(encryptionStage: CipherStage)(implicit mat: Materializer): Future[ByteString] =
    Source
      .single(secret)
      .via(encryptionStage)
      .runFold(ByteString.empty)(_ concat _)

  def toFileSecret(forFile: Path): DeviceFileSecret = {
    val filePath = forFile.toAbsolutePath.toString

    val salt = user.toBytes ++ device.toBytes ++ filePath.getBytes(StandardCharsets.UTF_8)

    val keyInfo = ByteString(s"${user.toString}-${device.toString}-$filePath-key")
    val ivInfo = ByteString(s"${user.toString}-${device.toString}-$filePath-iv")

    val hkdf = HKDF.fromHmacSha512()

    val prk = hkdf.extract(salt, secret.toArray)

    val key = hkdf.expand(prk, keyInfo.toArray, target.encryption.file.keySize)
    val iv = hkdf.expand(prk, ivInfo.toArray, target.encryption.file.ivSize)

    DeviceFileSecret(
      file = forFile,
      key = ByteString(key),
      iv = ByteString(iv)
    )
  }

  def toMetadataSecret(metadataCrate: Crate.Id): DeviceMetadataSecret = {
    val salt = user.toBytes ++ device.toBytes ++ metadataCrate.toBytes

    val keyInfo = ByteString(s"${user.toString}-${device.toString}-${metadataCrate.toString}-key")
    val ivInfo = ByteString(s"${user.toString}-${device.toString}-${metadataCrate.toString}-iv")

    val hkdf = HKDF.fromHmacSha512()

    val prk = hkdf.extract(salt, secret.toArray)

    val key = hkdf.expand(prk, keyInfo.toArray, target.encryption.metadata.keySize)
    val iv = hkdf.expand(prk, ivInfo.toArray, target.encryption.metadata.ivSize)

    DeviceMetadataSecret(
      key = ByteString(key),
      iv = ByteString(iv)
    )
  }
}

object DeviceSecret {
  def apply(
    user: User.Id,
    device: Device.Id,
    secret: ByteString
  )(implicit target: SecretsConfig): DeviceSecret =
    new DeviceSecret(user, device, secret)

  def decrypted(
    user: User.Id,
    device: Device.Id,
    encryptedSecret: ByteString,
    decryptionStage: CipherStage
  )(implicit mat: Materializer, target: SecretsConfig): Future[DeviceSecret] = {
    implicit val ec: ExecutionContext = mat.executionContext

    Source
      .single(encryptedSecret)
      .via(decryptionStage)
      .runFold(ByteString.empty)(_ concat _)
      .flatMap { rawDeviceSecret =>
        Future.successful(
          DeviceSecret(
            user = user,
            device = device,
            secret = rawDeviceSecret
          )
        )
      }
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/encryption/secrets/DeviceSecret.kt
package stasis.client_android.lib.encryption.secrets

import at.favre.lib.crypto.HKDF
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okio.Buffer
import okio.ByteString
import okio.ByteString.Companion.encodeUtf8
import okio.ByteString.Companion.toByteString
import okio.Source
import okio.buffer
import stasis.client_android.lib.model.core.CrateId
import stasis.client_android.lib.model.server.devices.DeviceId
import stasis.client_android.lib.model.server.users.UserId
import java.nio.file.Path

data class DeviceSecret(
    val user: UserId,
    val device: DeviceId,
    private val secret: ByteString,
    val target: Config
) : Secret() {
    suspend fun encrypted(encryptionStage: (Source) -> Source): ByteString =
        withContext(Dispatchers.IO) {
            encryptionStage(
                Buffer().write(secret)
            ).buffer().use { it.readByteString() }
        }

    fun toFileSecret(forFile: Path): DeviceFileSecret {
        val filePath = forFile.toAbsolutePath().toString()

        val salt = user.toBytes() + device.toBytes() + filePath.encodeToByteArray()

        val keyInfo = "$user-$device-$filePath-key".encodeUtf8()
        val ivInfo = "$user-$device-$filePath-iv".encodeUtf8()

        val hkdf = HKDF.fromHmacSha512()

        val prk = hkdf.extract(salt, secret.toByteArray())

        val key = hkdf.expand(prk, keyInfo.toByteArray(), target.encryption.file.keySize)
        val iv = hkdf.expand(prk, ivInfo.toByteArray(), target.encryption.file.ivSize)

        return DeviceFileSecret(
            file = forFile,
            key = key.toByteString(),
            iv = iv.toByteString()
        )
    }

    fun toMetadataSecret(metadataCrate: CrateId): DeviceMetadataSecret {
        val salt = user.toBytes() + device.toBytes() + metadataCrate.toBytes()

        val keyInfo = "$user-$device-$metadataCrate-key".encodeUtf8()
        val ivInfo = "$user-$device-$metadataCrate-iv".encodeUtf8()

        val hkdf = HKDF.fromHmacSha512()

        val prk = hkdf.extract(salt, secret.toByteArray())

        val key = hkdf.expand(prk, keyInfo.toByteArray(), target.encryption.metadata.keySize)
        val iv = hkdf.expand(prk, ivInfo.toByteArray(), target.encryption.metadata.ivSize)

        return DeviceMetadataSecret(
            key = key.toByteString(),
            iv = iv.toByteString()
        )
    }

    companion object {
        suspend fun decrypted(
            user: UserId,
            device: DeviceId,
            encryptedSecret: ByteString,
            decryptionStage: (Source) -> Source,
            target: Config
        ): DeviceSecret {
            val secret = withContext(Dispatchers.IO) {
                decryptionStage(
                    Buffer().write(encryptedSecret)
                ).buffer().use { it.readByteString() }
            }

            return DeviceSecret(
                user = user,
                device = device,
                secret = secret,
                target = target
            )
        }
    }
}
